gift-certificate-application/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,260 @@
+
+        if (!isValidGiftCertificate(giftCertificate)
+                || !isValidTags(giftCertificate.getTagList()) && giftCertificate.getTagList() != null) {
+            throw new FieldValidationException(INVALID_FIELDS_MSG);
+        }
+
+
+        if (isGiftCertificateExist(giftCertificate)) {
+            throw new GiftCertificateAlreadyExistException(String
+                    .format(GIFT_CERTIFICATE_ALREADY_EXIST_MSG, giftCertificate.getName()));
+        }
+
+        List<Tag> tagList = createTagsList(giftCertificate);
+        giftCertificate.setTagList(tagList);
+
+        return giftCertificateDao.create(giftCertificate);
+    }
+
+    private boolean isValidTags(List<Tag> tags) {
+        if (tags == null) {
+            return true;
+        }
+        return tags.stream().allMatch(tag -> tagValidator.isNameValid(tag.getName()));
+    }
+
+    private boolean isValidGiftCertificate(GiftCertificate giftCertificate) {
+        return giftCertificateValidator.validateAll(giftCertificate.getName(),
+                giftCertificate.getDescription(),
+                giftCertificate.getPrice(),
+                giftCertificate.getDuration());
+    }
+
+    private boolean isGiftCertificateExist(GiftCertificate giftCertificate) {
+        String giftCertificateName = giftCertificate.getName();
+        return giftCertificateDao.findByName(giftCertificateName).isPresent();
+    }
+
+    private List<Tag> createTagsList(GiftCertificate giftCertificate) {
+        if (giftCertificate.getTagList() == null) {
+            return Collections.emptyList();
+        }
+
+        List<Tag> tagList = new ArrayList<>();
+        for (Tag tag : giftCertificate.getTagList()) {
+            Optional<Tag> optionalTag = tagDao.findByName(tag.getName());
+            if (optionalTag.isPresent()) {
+                tag = optionalTag.get();
+            } else {
+                tag = tagDao.create(tag);
+            }
+            tagList.add(tag);
+        }
+
+        return tagList;
+    }
+
+    @Override
+    public List<GiftCertificate> findAll(CustomPagination pagination) {
+        Long gcNumber = giftCertificateDao.findEntitiesNumber(GiftCertificate.class);
+        pagination = paginationValidator.validatePagination(pagination, gcNumber);
+
+        return giftCertificateDao.findAll(pagination, GiftCertificate.class);
+    }
+
+    @Override
+    public GiftCertificate findByGiftCertificateId(Long id) {
+        return giftCertificateDao.findById(id, GiftCertificate.class)
+                .orElseThrow(() -> new GiftCertificateNotFoundException(String
+                .format(GIFT_CERTIFICATE_ID_NOT_FOUND_MSG, id)));
+    }
+
+    @Transactional
+    @Override
+    public void delete(Long id) {
+        Optional<GiftCertificate> optionalGiftCertificate = giftCertificateDao.findById(id, GiftCertificate.class);
+
+        if (!optionalGiftCertificate.isPresent()) {
+            String msg = String.format(GIFT_CERTIFICATE_ID_NOT_FOUND_MSG, id);
```

Comment:
Please name it as "message" instead of "msg", readability is important then compactness.